ooi_data_groom/plugins/botpt/botpt_plugin.py

```diff
@@ -168,29 +193,114 @@ def insert_precomputed_botpt_data(self, dataframe, metadata_record, computed_pro
+                                         & (dataframe.time <= last_to_delete_ntp)
+                compare_dataframe = dataframe[dataframe_compare_mask]
+                compare_size = len(compare_dataframe)
+
+                # compare the old and new data in the time range where they intersect
+                # only if they have the same shape in that time range
+                if compare_size == delete_size:
+                    log.info("dataframe to_delete before data comparison: size (%d), min (%d), max(%d)",
+                             delete_size, first_to_delete_ntp, last_to_delete_ntp)
+
+                    # mask to hold the results of the comparison of new and existing data
+                    comparison_result_mask = np.ones_like(to_delete.time, dtype=bool)
+
+                    # compare the values in each column using using numpy.isclose() with the default absolute (1e-08)
+                    # and relative (1e-05) tolerances which seem appropriate for the actual data in these streams
+                    for col in self.output_stream_derived_cols:
+                        comparison_result_mask = comparison_result_mask\
+                                               & np.isclose(to_delete[col], compare_dataframe[col], equal_nan=True)
+
+                    # index of all of the non-equal rows
+                    unequal_row_index = np.nonzero(np.logical_not(comparison_result_mask))[0]
+                    if unequal_row_index.size:
+                        # get the index of the first non-equal row
+                        first_unequal_row_index = unequal_row_index[0]
+                        # get the time of the first non-equal row. records in the existing and new dataframes
+                        # before this time are equal and do not need to be deleted and reinserted. These
+                        # rows of existing data can be simply be retained as they are.
+                        first_unequal_date = to_delete.time.iloc[first_unequal_row_index]
+                        log.info("current and new dataframe unequal_row_index size (%d), "
+                                 "first_unequal_row_index (%d), first_unequal_date (%d)",
+                                 unequal_row_index.size, first_unequal_row_index, first_unequal_date)
+
+                        # regenerate the dataframe exclude mask to exclude retained rows so that they are not
+                        # inserted again, which would create duplicate records
+                        dataframe_exclude_mask |= (dataframe.time >= to_delete.time.iloc[0])\
+                                               & (dataframe.time < first_unequal_date)
+
+                        # regenerate the delete mask to exclude retained records since we will not reinsert them
+                        # from the new dataframe
+                        delete_mask = (current_data.time >= first_unequal_date) & (current_data.time <= last_ntp)
```

Comment:
Could similar logic be performed for the last non-equal row - i.e. such and such is the last record/time that needs to change?

Reply:
Yes, that could be done but the benefits would be less pronounced. The main purpose of the change was to allow the process to run multiple times while the 3 hour bin of the source stream is actively coming into the system, without deleting the records from previous runs that have not changed during the current run. So the vast majority of the time, there won't be any unchanged records from the end of the current dataset because those records will be new data that came in since the previous run completed.